content/using_ec2_spot_instances_with_eks/jenkins/setup_agents.md

```diff
@@ -0,0 +1,60 @@
+---
+title: "Setting up Jenkins agents"
+date: 2018-08-07T08:30:11-07:00
+weight: 40
+---
+
+We now have our Jenkins Master running inside our EKS cluster, and we can reach the Jenkins dashboard via an ELB. We can create jobs which will be executed by Jenkins agents in pods within our cluster, but before we do that, let's create a dedicated Spot based nodegroup for our Jenkins agents, which will be slightly different from our existing nodegroups.
+
+#### Creating a new Spot Instances nodegroup for our Jenkins agent pods
+Earlier in the workshop, in the **Adding Spot Workers with eksctl** step, we created nodegroups that run a diversified set of Spot Instances to run our applications.
+
+Let's create a new eksctl nodegroup configuration file called `spot_nodegroup_jenkins.yml`. The Jenkins default resource requirements (Request and Limit CPU/Memory) are 512m (~0.5 vCPU) and 512Mi (~0.5 GB RAM), and since we are not going to perform any large build jobs in this workshop, we can stick to the defaults and also choose relatively small instance types that can accommodate the Jenkins agent pods.
+
+```
+cat <<EoF > ~/environment/spot_nodegroup_jenkins.yml
+apiVersion: eksctl.io/v1alpha5
+kind: ClusterConfig
+metadata:
+    name: eksworkshop-eksctl
+    region: $AWS_REGION
+nodeGroups:
+    - name: jenkins-agents-2vcpu-8gb-spot
+      minSize: 1
+      maxSize: 5
+      instancesDistribution:
+        instanceTypes: ["m5.large", "m5d.large", "m4.large","t3.large","t3a.large","m5a.large","t2.large"] 
+        onDemandBaseCapacity: 0
+        onDemandPercentageAboveBaseCapacity: 0
+        spotInstancePools: 4
+      labels:
+        lifecycle: Ec2Spot
+        intent: jenkins-agents
+      tags:
+          k8s.io/cluster-autoscaler/node-template/label/intent: jenkins-agents
```

Comment:
I'd suggest adding also the IAM parameter for cloudwatch as below. People that has completed the CloudWatch section will benefit from getting the logs for jenkins on CloudWatch

```
      iam:
        withAddonPolicies:
          cloudWatch: true
```

Reply:
Ok, Interesting, after making the change to cluster autoscaler, the pod for cluster autoscaler ended up in the jenkins agent. Given the jenkins is missing the following policy for getting access to asg, 

```
      iam:
        withAddonPolicies:
          cloudWatch: true
          autoScaler: true
```

I started getting this errors 

```
E1024 00:11:44.387500       1 aws_manager.go:148] Failed to regenerate ASG cache: AccessDenied: User: arn:aws:sts::781623751140:assumed-role/eksctl-eksworkshop-eksctl-nodegro-NodeInstanceRole-2KJFOTZZYN6H/i-0528c20938e4d347f is not authorized to perform: autoscaling:DescribeAutoScalingGroups
        status code: 403, request id: dc9cb901-f5f2-11e9-aadd-edf391e9ed24
F1024 00:11:44.387966       1 aws_cloud_provider.go:335] Failed to create AWS Manager: AccessDenied: User: arn:aws:sts::781623751140:assumed-role/eksctl-eksworkshop-eksctl-nodegro-NodeInstanceRole-2KJFOTZZYN6H/i-0528c20938e4d347f is not authorized to perform: autoscaling:DescribeAutoScalingGroups
```

Solution is either to add the `autoscaler` policy to all the nodegroups, or to change the configuration in the autoscaler template file and add a nodeSelector section to select nodes with intent label set to control-apps.

Reply:
I reckon the second option is more prescriptive ,adding the nodeSelector in the cluster autoscaler config.

Reply:
second option done. I did not add these different capabilities to the Jenkins nodegroup on purpose - it is dedicated for Jenkins workers and should not run other workloads. agreed?